### PR TITLE
Document the nix-util Meson build system more extensively

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

I hope this will make it easier to maintain, and also make it easier for others to assist with porting the rest of the build system to Meson.

# Context

https://github.com/NixOS/nix/issues/2503

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -243,6 +243,7 @@
           changelog-d;
         default = self.packages.${system}.nix;
       } // lib.concatMapAttrs
+        # We need to "flatten" packages we care about to pass `flake check`.
```
Comment (roberth) [inline flake.nix:246]:
```suggestion
        # We need to "flatten" package sets to pass `flake check`.
```

- sets
- what if we don't care?

File: src/libutil/meson.build [hunk 2]
```diff
@@ -53,7 +80,8 @@ boost = dependency(
   'boost',
   modules : ['context', 'coroutine'],
 )
-# Actually public, but wrong type of dep for pkg config
+# This is public, but not a pkg-config dependency unfortunately, so we
```
Comment (roberth) [inline src/libutil/meson.build:83]:
```suggestion
# boost is a public dependency, but not a pkg-config dependency unfortunately, so we
```

Explicit references are preferable, as they don't bitrot when new items are added.

**Review by roberth — COMMENTED** [review]
I've (quickly) proofread the docs.
I'm curious what the change near the bottom of `libutil/meson.build` is about. It seems to be a real change, not just docs; was that intentional?

File: src/libutil/meson.build [hunk 3]
```diff
@@ -8,25 +8,50 @@ project('nix-util', 'cpp',
...
+# are thus part of the *implementation* of the library, but not its
+# *interface*.
+#
+# See `man pkg-config` for some details.
 deps_private = [ ]
+
+# These are public dependencies with pkg-config files. Public is the
+# opposite of private: these dependencies are used in installed header
+# files. They are part of the interface (and implementation) of the
+# library.
```
Comment (roberth) [inline src/libutil/meson.build:30]:
```suggestion
# library.
# This concept is mostly unrelated to our own concept of a public (stable) API, for consumption
# outside of the Nix repository.
```

File: flake.nix (hunk 1 above)
Comment (Ericson2314) [reply flake.nix:246]: I meant we're not doing all of `nixpkgsFor.*` but filtering down to a select few packages. I don't like my phrasing either, is there a better way?

Comment (fricklerhandwerk) [reply flake.nix:246]:
```suggestion
        # We need to flatten recursive attribute sets of derivations to pass `flake check`.
```

**Review by fricklerhandwerk — APPROVED** [review]
Nice

**Pull request merged**